core/src/main/java/org/elasticsearch/discovery/DiscoveryModule.java

```diff
@@ -62,16 +57,30 @@ public DiscoveryModule(Settings settings, TransportService transportService, Net
         addDiscoveryType("none", NoneDiscovery.class);
         addDiscoveryType("zen", ZenDiscovery.class);
 
-        Map<String, Supplier<UnicastHostsProvider>> hostProviders = new HashMap<>();
-        hostProviders.put("zen", () -> Collections::emptyList);
-        for (DiscoveryPlugin plugin : plugins) {
-            plugin.getZenHostsProviders(transportService, networkService).entrySet().forEach(entry -> {
-                if (hostProviders.put(entry.getKey(), entry.getValue()) != null) {
-                    throw new IllegalArgumentException("Cannot specify zen hosts provider [" + entry.getKey() + "] twice");
-                }
-            });
+        String discoveryType = DISCOVERY_TYPE_SETTING.get(settings);
+        if (discoveryType.equals("none") == false) {
+            Map<String, Supplier<UnicastHostsProvider>> hostProviders = new HashMap<>();
+            hostProviders.put("zen", () -> Collections::emptyList);
+            for (DiscoveryPlugin plugin : plugins) {
+                plugin.getZenHostsProviders(transportService, networkService).entrySet().forEach(entry -> {
```

Comment:
I think long term that might be better, but what I have here follows what we have done with other extension points. Long term I think I might like to still have a selection, but have it be automatically wired to the name of the plugin.